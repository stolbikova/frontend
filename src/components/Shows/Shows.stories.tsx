import * as React from "react";
import { storiesOf } from "@storybook/react";
import {Shows} from "./Shows";
import {BrowserRouter as Router} from 'react-router-dom';

storiesOf("ShowsComponent", module).add("test render", () => (
  <Router>
    <Shows />
  </Router>
));
