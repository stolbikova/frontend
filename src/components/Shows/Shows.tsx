import * as React from 'react';
import styled, {withProps} from "styled";
import {Show} from '../Show/Show';
import {ShowPreview} from "../ShowPreview/ShowPreview";
import {ShowI} from '../../types';
import {API_URL} from "../../constants";
import {useNavigate, useLocation} from "react-router-dom";

interface ShowsProps {
  showsApi?: string;
}

const Wrapper = withProps()(styled.main)`
    display: flex;
    flex-direction: column;
    align-items: center;

    ${props => props.theme.media.min.tablet`
      flex-direction: row;
    `}
`;
const ShowPreviewWrapper = styled.ul`
   display: flex;
   margin-top: 10px;
   flex-wrap: wrap;
   align-items: center;
   justify-content: center;
   list-style-type: none;
   padding: 0;

   ${props => props.theme.media.min.tablet`
      flex-direction: column;
      margin-left: 10px;
      margin-top: 30px;
   `}
`;

export const Shows: React.FC<ShowsProps> = ({showsApi}) => {
  const [data, setData] = React.useState<ShowI []>([]);
  const [selected, setSelected] = React.useState<ShowI | null>(null);
  const [error, setError] = React.useState();
  const navigate = useNavigate();
  const location = useLocation();
  const setSelectedShow = (showId: string) => {
    const newSelected = data.find(s => s.id === showId);
    if (!newSelected) return;
    setSelected(newSelected);
  }
  React.useEffect(() => {
    if (showsApi) fetch(`${API_URL}/shows`)
      .then(response => response.json())
      .then(json => setData(json))
      .catch(err => {
        setError(err);
      })
  }, [showsApi, setError, setData]);
  React.useEffect(() => {
    if (data && data[0]) {
      if (location && location.search.includes('show_id=')) {
        const showId = location.search.split('show_id=')[1];
        setSelectedShow(showId);
      } else {
        setSelected(data[0]);
      }
    }
  }, [data]);
  React.useEffect(() => {
    if (location && location.search.includes('show_id=')) {
      const showId = location.search.split('show_id=')[1];
      setSelectedShow(showId);
    }
  }, [location]);

  const handleChangeShow = (id: string) => {
    setSelectedShow(id);

    const nextURL = `?show_id=${id}`;
    navigate(nextURL);
  }

  if (error) {
    return (
      <div>
        <span>Data fetching error. Please start the API server.</span>
      </div>)
  }

  if (!data) {
    return null;
  }

  return (
    <Wrapper>
      {selected ? <Show {...selected} /> : null}
      <nav>
        <ShowPreviewWrapper style={{display: 'flex'}}>
          {data.map((s, idx) => (
            <ShowPreview key={idx} {...s} onClick={handleChangeShow} selected={!!selected && (s.id === selected.id)}/>
          ))}
        </ShowPreviewWrapper>
      </nav>
    </Wrapper>
  )
}
