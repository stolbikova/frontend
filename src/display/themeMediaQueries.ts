/**
 * Creates an object with the keys "min" and "max" which each have keys for the
 * devices (viewport sizes) listed by the "sizes" object. Each viewport function
 * is a template string function which creates a media query.
 *
 * @see https://www.styled-components.com/docs/advanced#media-templates
 * @example
 * ```
 *  const Content = styled.div`
 *    height: 3em;
 *
 *    ${props => props.theme.media.min.tablet`
 *      background: dodgerblue;
 *    `}
 *  `;
 * ```
 */
import {
  Interpolation,
  ThemedStyledProps,
  FlattenInterpolation,
} from "styled-components";
import { css } from "styled";
import { Theme } from "./theme";

// Add additional media queries here.
const sizes: Record<DeviceType, number> = {
  tablet: 980,
};

const themeMediaQueries = { min: {}, max: {} } as ThemeMediaQueries;

Object.keys(sizes).reduce((acc, label) => {
  const device = label as DeviceType;

  (["min", "max"] as Range[]).forEach(range => {
    const mediaTemplate: ThemedCssFunction = (
      strings,
      ...interpolations
    ) => css`
      @media (${range}-width: ${sizes[device]}px) {
        ${css(strings, ...interpolations)};
      }
    `;

    acc[range][device] = mediaTemplate;
  });

  return acc;
}, themeMediaQueries);

export default themeMediaQueries;

export type ThemeMediaQueries = Record<
  Range,
  Record<DeviceType, ThemedCssFunction>
>;

// Pulled from styled-components type definitions.
type ThemedCssFunction = (
  strings: TemplateStringsArray,
  ...interpolations: Interpolation<
    ThemedStyledProps<{}, Theme & { media: ThemeMediaQueries }>
  >[]
) => FlattenInterpolation<
  ThemedStyledProps<{}, Theme & { media: ThemeMediaQueries }>
>[];

type Range = "min" | "max";
type DeviceType = "tablet";
