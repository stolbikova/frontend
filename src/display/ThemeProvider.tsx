import * as React from "react";
import {
    injectGlobal,
    ThemeProvider as StyledComponentsThemeProvider,
} from "styled";
import theme from "./theme";
import themeMediaQueries from "./themeMediaQueries";

const ThemeProvider: React.FC<{}> = props => (
    <StyledComponentsThemeProvider theme={{ ...theme, media: themeMediaQueries }}>
        <>
            {/* react-helmet, ... */}
            {props.children}
        </>
    </StyledComponentsThemeProvider>
);

/* tslint:disable-next-line:no-unused-expression */
injectGlobal`
  html {
    box-sizing: border-box;
  }

  *, *:before, *:after {
    box-sizing: inherit;
  }
`;

export default ThemeProvider;
