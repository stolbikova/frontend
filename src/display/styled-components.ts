// This module is imported using the import:
// import styled from "styled";
// Reference webpack.config.ts / jest.config.js / tsconfig.json

// https://www.styled-components.com/docs/api#typescript
// tslint:disable no-duplicate-imports
import * as styledComponents from "styled-components";
import { ThemedStyledComponentsModule } from "styled-components";
import { Theme } from "./theme";
import { ThemeMediaQueries } from "./themeMediaQueries";

const {
  default: styled,
  css,
  injectGlobal,
  ThemeProvider,
} = styledComponents as ThemedStyledComponentsModule<
  Theme & { media: ThemeMediaQueries }
>;

type ThemedStyledFunction<T> = styledComponents.ThemedStyledFunction<
  T,
  Theme & { media: ThemeMediaQueries }
>;

const withProps = <U>() => <P>(fn: ThemedStyledFunction<P>) =>
  fn as ThemedStyledFunction<P & U>;

export { css, injectGlobal, withProps, ThemeProvider };
export default styled;
