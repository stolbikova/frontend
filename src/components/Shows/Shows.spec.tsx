import * as React from "react";
import * as renderer from "react-test-renderer";
import {BrowserRouter as Router} from 'react-router-dom';
import {Shows} from "./Shows";
import {ThemeProvider} from "../../display";

const createRenderer = () =>
  renderer.create(
    <Router>
      <ThemeProvider>
      <Shows />
      </ThemeProvider>
    </Router>,
  );

it("renders Shows subcomponent", () => {
  const component = createRenderer();
  expect(component.toJSON()).toMatchSnapshot();
});
