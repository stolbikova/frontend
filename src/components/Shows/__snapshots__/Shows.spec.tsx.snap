// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Shows subcomponent 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-top: 10px;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  list-style-type: none;
  padding: 0;
}

@media (min-width:980px) {
  .c0 {
    -webkit-flex-direction: row;
    -ms-flex-direction: row;
    flex-direction: row;
  }
}

@media (min-width:980px) {
  .c1 {
    -webkit-flex-direction: column;
    -ms-flex-direction: column;
    flex-direction: column;
    margin-left: 10px;
    margin-top: 30px;
  }
}

<main
  className="c0"
>
  <nav>
    <ul
      className="c1"
      style={
        Object {
          "display": "flex",
        }
      }
    />
  </nav>
</main>
`;
